wallet/src/de/schildbach/wallet/service/platform/TopUpRepository.kt

```diff
@@ -91,25 +92,27 @@ class TopUpRepositoryImpl @Inject constructor(
 ) : TopUpRepository {
     companion object {
         private val log = LoggerFactory.getLogger(TopUpRepositoryImpl::class.java)
+        private const val MIN_DUST_FACTOR = 10L
     }
 
     private val platform = platformRepo.platform
     private val authExtension by lazy { walletDataProvider.wallet!!.getKeyChainExtension(AuthenticationGroupExtension.EXTENSION_ID) as AuthenticationGroupExtension }
 
-    override fun createAssetLockTransaction(
+    override suspend fun createAssetLockTransaction(
         blockchainIdentity: BlockchainIdentity,
         username: String,
         keyParameter: KeyParameter?,
         useCoinJoin: Boolean
     ) {
-        Context.propagate(walletDataProvider.wallet!!.context)
         val fee = if (Names.isUsernameContestable(username)) {
             Constants.DASH_PAY_FEE_CONTESTED
         } else {
             Constants.DASH_PAY_FEE
         }
-        val balance = walletDataProvider.wallet!!.getBalance(Wallet.BalanceType.ESTIMATED_SPENDABLE)
-        val emptyWallet = balance == fee && balance <= (fee + Transaction.MIN_NONDUST_OUTPUT)
+        val balance = walletDataProvider.observeSpendableBalance().first()
+        val emptyWallet = balance == fee ||
+                (balance >= fee && balance <= (fee + Transaction.MIN_NONDUST_OUTPUT.multiply(MIN_DUST_FACTOR)))
+        Context.propagate(walletDataProvider.wallet!!.context)
```

Comment:
If a user has about 0.032 DASH and then mixes them to get just above 0.03 DASH, this code wouldn't set emptyWallet to `true`, then the username creation process would get stuck and report an insufficient funds exception in the logs.  the only way to recover was to deposit more DASH and mix a bit more so that the fee could be covered.

Our goal is to empty the wallet if the username creation fee is just a bit higher than the requirement (0.03 in this case) to avoid such errors.

In this case, we were looking at the wrong balance (total) instead of spendable.

There still might be a way for this to fail.